pallets/oracle-data-collection/src/tests.rs

```diff
@@ -0,0 +1,309 @@
+			}
+		});
+		MockIsAdmin::mock_check(|(admin, collection_id)| {
+			assert_eq!(admin, ADMIN);
+			assert_eq!(collection_id, COLLECTION_ID);
+			true
+		});
+	}
+
+	pub fn prepare_provider() {
+		MockProvider::mock_get(|(account, collection_id), key| {
+			assert_eq!(collection_id, &COLLECTION_ID);
+			match (account, key) {
+				(&FEEDER_1, &KEY_A) => Ok(Some((100, 50))),
+				(&FEEDER_2, &KEY_A) => Ok(Some((101, 45))),
+				(&FEEDER_3, &KEY_A) => Ok(Some((102, 55))),
+				(&FEEDER_1, &KEY_B) => Ok(Some((1000, 500))),
+				(&FEEDER_2, &KEY_B) => Ok(None),
+				(&FEEDER_3, &KEY_B) => Ok(None),
+				(&FEEDER_1, &KEY_ERR) => Err(DispatchError::Other("get err")),
+				(&FEEDER_2, &KEY_ERR) => Err(DispatchError::Other("get err")),
+				(&FEEDER_3, &KEY_ERR) => Err(DispatchError::Other("get err")),
+				(&FEEDER_1, &KEY_NONE) => Ok(None),
+				(&FEEDER_2, &KEY_NONE) => Ok(None),
+				(&FEEDER_3, &KEY_NONE) => Ok(None),
+				_ => unreachable!(),
+			}
+		});
+	}
+}
+
+mod util {
+	use super::*;
+
+	pub fn update_feeders(key: OracleKey, feeders: Vec<AccountId>) {
+		let feeders: BoundedVec<_, _> = feeders.try_into().unwrap();
+
+		MockChangeGuard::mock_note(|_, _| Ok(CHANGE_ID));
+		MockChangeGuard::mock_released({
+			let feeders = feeders.clone();
+			move |_, _| Ok(Change::Feeders(key, feeders.clone()))
+		});
+		MockIsAdmin::mock_check(|_| true);
+
+		OracleCollection::propose_update_feeders(
+			RuntimeOrigin::signed(ADMIN),
+			COLLECTION_ID,
+			key,
+			feeders,
+		)
+		.unwrap();
+
+		OracleCollection::apply_update_feeders(
+			RuntimeOrigin::signed(ADMIN),
+			COLLECTION_ID,
+			CHANGE_ID,
+		)
+		.unwrap();
+
+		MockChangeGuard::mock_note(|_, _| panic!("no note() mock"));
+		MockChangeGuard::mock_released(|_, _| panic!("no released() mock"));
+		MockIsAdmin::mock_check(|_| panic!("no check() mock"));
+	}
+}
+
+#[test]
+fn updating_feeders() {
+	new_test_ext().execute_with(|| {
+		let feeders: BoundedVec<_, _> = vec![FEEDER_1, FEEDER_2].try_into().unwrap();
+
+		mock::prepare_update_feeders(KEY_A, &feeders);
+
+		assert_ok!(OracleCollection::propose_update_feeders(
+			RuntimeOrigin::signed(ADMIN),
+			COLLECTION_ID,
+			KEY_A,
+			feeders
+		));
+
+		assert_ok!(OracleCollection::apply_update_feeders(
+			RuntimeOrigin::signed(ADMIN),
+			COLLECTION_ID,
+			CHANGE_ID,
+		));
+	});
+}
+
+#[test]
+fn updating_feeders_wrong_admin() {
+	new_test_ext().execute_with(|| {
+		let feeders: BoundedVec<_, _> = vec![FEEDER_1, FEEDER_2].try_into().unwrap();
+
+		mock::prepare_update_feeders(KEY_A, &feeders);
+		MockIsAdmin::mock_check(|_| false);
+
+		assert_err!(
+			OracleCollection::propose_update_feeders(
+				RuntimeOrigin::signed(ADMIN),
+				COLLECTION_ID,
+				KEY_A,
+				feeders
+			),
+			Error::<Runtime>::IsNotAdmin
+		);
+	});
+}
+
+#[test]
+fn register() {
+	new_test_ext().execute_with(|| {
+		assert_ok!(OracleCollection::register_id(&KEY_A, &COLLECTION_ID));
+		assert_eq!(Keys::<Runtime>::get(COLLECTION_ID, KEY_A).usage_refs, 1);
```

Comment:
Nit: Could add check for `KeysAdded` event dispatch

Reply:
Done